src/Control/Monad/Cofree/Trans.purs

```diff
@@ -0,0 +1,156 @@
+tail = map T.snd <<< runCofreeT
+
+-- Note: This cannot be automatically derived because 'a' also appears in the
+-- 'fst' position of the inner Tuple.
+instance functorCofreeT :: (Functor m, Functor f) => Functor (CofreeT f m) where
+  map f (CofreeT inner) = CofreeT $ map (map (bimap f (map (map f)))) inner
+
+instance applyCofreeT :: (Apply m, Apply f) => Apply (CofreeT f m) where
+  apply (CofreeT innerF) (CofreeT inner) =
+    CofreeT
+      $ \_ ->
+        go <$> innerF unit <*> inner unit
+    where
+      go (Tuple f nextF) (Tuple x nextX) =
+        Tuple (f x) (lift2 (<*>) nextF nextX)
+
+instance applicativeCofreeT :: (Applicative m, Apply f, Plus f) => Applicative (CofreeT f m) where
+  pure a = CofreeT $ \_ -> pure (Tuple a empty)
+
+instance bindCofreeT :: (Monad m, Alt f, Apply f) => Bind (CofreeT f m) where
+  bind (CofreeT inner) f =
+    CofreeT
+      $ \_ -> do
+        (Tuple a m) <- inner unit
+        let (CofreeT next) = f a
+        (Tuple b n) <- next unit
+        pure $ Tuple b (n <|> map (_ >>= f) m)
+
+instance monadCofreeT :: (Monad m, Plus f, Apply f) => Monad (CofreeT f m)
+
+instance monadTransCofreeT :: Plus f => MonadTrans (CofreeT f) where
+  lift = cofreeT' <<< map go
+    where
+      go x = Tuple x empty
+
+instance monadEffectCofreeT :: (MonadEffect m, Plus f, Apply f) => MonadEffect (CofreeT f m) where
+  liftEffect eff = cofreeT' $ go <$> liftEffect eff
+    where
+      go a = Tuple a empty
+
+instance monadAffCofreeT :: (MonadAff m, Plus f, Apply f) => MonadAff (CofreeT f m) where
+  liftAff aff = cofreeT' $ go <$> liftAff aff
+    where
+      go a = Tuple a empty
+
+instance comonadCofreeCofreeT :: (Comonad m, Functor f) => ComonadCofree f (CofreeT f m) where
+  unwrapCofree = extract <<< tail
+
+instance comonadTransCofreeT :: ComonadTrans (CofreeT f) where
+  lower = head
+
+instance comonadAskCofreeT :: (Functor f, ComonadAsk e m) => ComonadAsk e (CofreeT f m) where
+  ask = ask <<< tail
+
+instance foldableCofreeT :: (Foldable m, Foldable f) => Foldable (CofreeT f m) where
+  foldMap f (CofreeT inner) = foldMap go $ inner unit
+    where
+      go (Tuple a next) = f a <> foldMap (foldMap f) next
+
+  foldr abb b = foldrDefault abb b
+
+  foldl bab b = foldlDefault bab b
+
+instance traversableCofreeT :: (Traversable m, Traversable f) => Traversable (CofreeT f m) where
+  traverse f (CofreeT inner) =
+    cofreeT' <$> traverse go (inner unit)
+    where
+      go (Tuple a next) = Tuple <$> f a <*> traverse (traverse f) next
+
+  sequence = sequenceDefault
+
+instance extendCofreeT :: (Comonad m, Functor f) => Extend (CofreeT f m) where
+  extend f (CofreeT inner) = CofreeT $ \_ -> extend go (inner unit)
+    where
+      go w = Tuple (f $ cofreeT' w) $ extend f <$> T.snd (extract w)
+
+instance comonadCofreeT :: (Comonad m, Functor f) => Comonad (CofreeT f m) where
+  extract = extract <<< head
+
+instance eqCofreeT :: Eq (m (Tuple a (f (CofreeT f m a)))) => Eq (CofreeT f m a) where
```

Comment:
This won't be guaranteed to terminate in most actual use cases. Is there a reason for adding it?

Reply:
Is it because we expect `CofreeT`s to generally be infinite? The main reason I added it is because this instance exists in Haskell as well, but I see your point. I'm not sure this would get a lot of (legitimate) use. Maybe it's better to remove and consider re-adding if we get requests/issues for real use cases.

Reply:
Would it be possible to use `Eq1 m` instead of this constraint, if we keep it?

Reply:
I removed both `Eq` and `Ord` for now. It might be possible to use `Eq1` though, so I'll keep that in mind if we end up implementing it eventually.